src/v/v8_engine/executor.h

```diff
@@ -0,0 +1,225 @@
+    // complete task
+    ss::future<> get_future() {
+        return _on_done.wait().discard_result().then([this] {
+            if (_exception) {
+                return ss::make_exception_future(_exception);
+            } else {
+                return ss::now();
+            }
+        });
+    }
+
+    Func _func;
+    seastar::readable_eventfd _on_done;
+
+    std::chrono::milliseconds _timeout;
+};
+
+// This class implement queue for submit task from seastar to std::thread. Only
+// one core with seastar can submit task to queue. Only one std::thread can
+// consume task from queue.
+class spsc_queue {
+public:
+    explicit spsc_queue(size_t queue_size);
+
+    // Close queue and break waiting on seastar mutex
+    void close();
+
+    // Push new item to queue. Only one future can wait on readable_eventfd
+    ss::future<> push(work_item* item);
+
+    // Pop element from queue (blocking)
+    work_item* pop();
+
+    bool empty();
+
+private:
+    mutex _push_mutex;
+    seastar::readable_eventfd _is_not_full;
+
+    std::atomic<bool> is_stopped{false};
+
+    boost::lockfree::spsc_queue<work_item*> _items;
+
+    std::mutex _std_mutex;
+    std::condition_variable _has_element_cv;
+    static constexpr std::chrono::milliseconds _timeout_cond_wait_std_thread_ms{
+      30};
+
+    ss::shard_id _ss_mutex_shard;
+};
+
+} // namespace internal
+
+// This class implement executor (std::thread) for runing v8 script. It
+// uses boost::spsc_queue for storing task.
+// For use we need create ss::shared<executor> and run submit like
+// executor_sharded.invoke_on(<core with executor>)
+class executor {
+public:
+    executor(uint8_t cpu_id, size_t queue_size);
+
+    executor(const executor& other) = delete;
+    executor& operator=(const executor& other) = delete;
+    executor(executor&& other) = delete;
+    executor& operator=(executor&& other) = delete;
+
+    ~executor();
+
+    // Stop executor. Stop queue with task and close gate
+    ss::future<> stop();
+
+    bool is_stopping() const;
+
+    /// Submit new task in executor.
+    ///
+    /// \param start_func is used for set watchdog
+    /// \param func_for_executor is v8 script
+
+    // clang-format off
+    template<typename WrapperFuncForExecutor>
+    CONCEPT(requires requires(WrapperFuncForExecutor func) {
+        { func() } -> std::same_as<void>;
+        { func.cancel() } -> std::same_as<void>;
+        { func.on_timeout() } -> std::same_as<void>;
+    })
+    // clang-format on
+
+    ss::future<> submit(
+      WrapperFuncForExecutor&& func_for_executor,
+      std::chrono::milliseconds timeout) {
+        gate_guard guard{_gate};
+
+        auto new_task
+          = std::make_unique<internal::task<WrapperFuncForExecutor>>(
+            std::forward<WrapperFuncForExecutor>(func_for_executor), timeout);
+
+        co_await _tasks.push(new_task.get());
+        co_await std::move(new_task->get_future());
```

Comment:
I've encountered a few cases where co_await std::move(f) was needed, but this doesn't look like one of those cases. did you try without std::move?

Reply:
Hm, I got some problems without std::move
But I think it was from another bug, now it works